### PR TITLE
Modified EulerAngleProvider2RGBAux to work when regions are not assigned to a grain

### LINKED ISSUE #7384: Fix EulerAngleProvider2RGB to work with points with no grain ID
### Description of the enhancement or error report
We need to modify EulerAngleProvider2RGB to output white or black when there is no grain ID

### Rationale for the enhancement or information for reproducing the error
Currently EulerAngleProvider2RGB errors out when there is no grain ID provided

### Identified impact
This will change EulerAngleProvider2RGB, but no tests should need changing.

### PR DESCRIPTION
**Pull request by tonkmr**
Modified the auxkernel and added a test that tests the new behavior. Closes #7384

### CODE REVIEW

**Comment by permcody** [comment]
Interesting - A mismatch that only showed up in debug?! I guess this is why I need more data files. I'm really not happy with that. I'll see if I can pull this down and take a look.

**Comment by tonkmr** [comment]
I had to use a large number of order parameters to make this work as well (25). I decreased the resolution of a larger data set to make the test, so there are a number of very small grains.

**Comment by permcody** [comment]
I added the "advanced order parameter" remapping algorithm to the existing
111 grain simulation and got it down to 6 order parameters! I haven't
checked that code in yet though because I'm still working a few issues out
with Daniel. Anyway, I'll try it on this newer dataset

**Comment by tonkmr** [comment]
I made a new data set in which I crop a smaller region rather than decreasing the resolution and it results in a better data file. In the current one there are one pixel grains and I just don't think that is a good test problem. So, I am going to switch. I can send you the original data file that required 20 order parameters.

**Comment by permcody** [comment]
One pixel grains may not show up since they may not contribute enough to
the elemental dof where we sample.

**Comment by tonkmr** [comment]
I am still having the same failure in debug, even with the cropped data set without single pixel grains.

**Comment by tonkmr** [comment]
@permcody I would appreciate help with this. This runs fine in debug on my machine. Any thoughts?

**Comment by permcody** [comment]
I still haven't had any time (in the last two days) to look at it. I'll see
if I can get to it soon.

**Comment by permcody** [comment]
I just pushed up an important bug fix here:
https://github.com/idaholab/moose/pull/7393/commits/5748929bd435e9e25ef9b7b0cd0696085c8e97ca

There was a problem with flooding when using the connecting threshold. It's likely related to this problem. I haven't tested yet but these two grain texture tests were the ones that diff'd with the patch.

File: modules/phase_field/src/action/EulerAngle2RGBAction.C [hunk 1]
```diff
@@ -22,6 +22,7 @@ InputParameters validParams<EulerAngle2RGBAction>()
   params.addRequiredParam<MooseEnum>("crystal_structure",structure_enum,"Crystal structure of the material");
   params.addRequiredParam<UserObjectName>("euler_angle_provider", "Name of Euler angle provider user object");
   params.addRequiredParam<UserObjectName>("grain_tracker_object", "The GrainTracker UserObject to get values from.");
+  params.addParam<std::vector<Real> >("no_grain_color", std::vector<Real>{ 0, 0 , 0 }, "RGB value of color used to represent area with no grains");
```
Comment (dschwen) [inline modules/phase_field/src/action/EulerAngle2RGBAction.C]: Could be range checked with `"no_grain_color_size = 3"`

File: modules/phase_field/src/auxkernels/EulerAngleProvider2RGBAux.C [hunk 2]
```diff
@@ -31,21 +32,29 @@ EulerAngleProvider2RGBAux::EulerAngleProvider2RGBAux(const InputParameters & par
...
+  if (grain_id >= 0)
+    angles = _euler.getEulerAngles(grain_id);
 
-  // Return RGB value computed from Euler angles
-  Point RGB = euler2RGB(_sd, angles(0) / 180.0 * libMesh::pi, angles(1) / 180.0 * libMesh::pi, angles(2) / 180.0 * libMesh::pi, 1.0, _xtal_class);
+  // Return RGB value computed from Euler
+  Point RGB;
+  if (grain_id < 0) //If not in a grain, return white
+      for (unsigned int i = 0; i < 3; ++i)
+        RGB(i) = _no_grain_color[i];
```
Comment (dschwen) [inline modules/phase_field/src/auxkernels/EulerAngleProvider2RGBAux.C]: `_no_grain_color` could be shorter than 3 (see above)

File: modules/phase_field/src/action/EulerAngle2RGBAction.C (hunk 1 above)
Comment (permcody) [inline modules/phase_field/src/action/EulerAngle2RGBAction.C]: Let's not do this at all. I suspect that my newly identified bug will take care of this.

File: modules/phase_field/src/auxkernels/EulerAngleProvider2RGBAux.C [hunk 3]
```diff
@@ -22,6 +22,7 @@ InputParameters validParams<EulerAngleProvider2RGBAux>()
   params.addParam<MooseEnum>("output_type", output_types, "Type of value that will be outputted");
   params.addRequiredParam<UserObjectName>("euler_angle_provider", "Name of Euler angle provider user object");
   params.addRequiredParam<UserObjectName>("grain_tracker_object", "The GrainTracker UserObject to get values from.");
+  params.addParam<std::vector<Real> >("no_grain_color", std::vector<Real>{ 0, 0 , 0 }, "RGB value of color used to represent area with no grains");
```
Comment (dschwen) [inline modules/phase_field/src/auxkernels/EulerAngleProvider2RGBAux.C]: alternatively make this of type `Point` (then you can assign it without a loop)

File: modules/phase_field/src/auxkernels/EulerAngleProvider2RGBAux.C [hunk 4]
```diff
@@ -31,21 +32,29 @@ EulerAngleProvider2RGBAux::EulerAngleProvider2RGBAux(const InputParameters & par
...
-  const RealVectorValue angles = _euler.getEulerAngles(grain_id);
+  RealVectorValue angles;
+  if (grain_id >= 0)
+    angles = _euler.getEulerAngles(grain_id);
 
-  // Return RGB value computed from Euler angles
-  Point RGB = euler2RGB(_sd, angles(0) / 180.0 * libMesh::pi, angles(1) / 180.0 * libMesh::pi, angles(2) / 180.0 * libMesh::pi, 1.0, _xtal_class);
+  // Return RGB value computed from Euler
+  Point RGB;
+  if (grain_id < 0) //If not in a grain, return white
```
Comment (dschwen) [inline modules/phase_field/src/auxkernels/EulerAngleProvider2RGBAux.C]: comment is outdated

File: modules/phase_field/src/action/EulerAngle2RGBAction.C (hunk 1 above)
Comment (dschwen) [inline modules/phase_field/src/action/EulerAngle2RGBAction.C]: No. This is a required feature. Tracked grains are not at all required to cover the entire domain.

**Comment by tonkmr** [comment]
Once #7393 is merged, we can see if it fixes my problem.

**Comment by tonkmr** [comment]
@permcody I have run this again with the latest version of the code but I am still getting the same error "EBSD centers: 56 Grain Tracker Centers: 55" but only in debug. Any help would be appreciated. @frombs needs this capability for a paper he is writing right now.

**Comment by permcody** [comment]
OK - I'll check it out.

**Comment by tonkmr** [comment]
It runs fine in debug on my machine, by the way.

**Comment by permcody** [comment]
I looked into this. It's related to the same issue that Brad and I (mostly Brad) discovered on the 111 grain dataset we have. This data set contains a non contiguous grain. See screenshot:

bad_ebsd

I added that Grain to EBSD map, but after talking to @dschwen - it's not enough. These grains really should be part of the same feature even if they are disconnected like this. Technically they should be allowed to coalesce under the right driving force. Right now what's happening is that the GrainTracker is adding one extra grain to deal with this disconnected piece which may cause you a headache.

I guess if you want to skip this issue for now you could just crop the data set a bit more to remove this bad boy:

**Comment by tonkmr** [comment]
@permcody thanks for looking into this. Any idea on why this is only showing up in debug?

**Comment by permcody** [comment]
I see the warning in all modes (mismatched centers). The problem (if you are seeing one) is that the grain tracker will create a number for that extra piece (i.e. 56) and if you use that number to index into any data structures or to ask the EBSD reader for information, you are doing something bad and it might just do something undefined in opt mode but I guess we have the right asserts or extra checks from preventing that from happening in debug mode.

**Comment by frombs** [comment]
@permcody I thought you added an option to GrainTracker last year that uses the grain (feature) numbering from EBSDReader as input when creating the grain maps.  Is that option still available?

**Comment by permcody** [comment]
Yes, it's on automatically. If you turn on the UNIQUE_REGION field you'll see it only indexes up to 55 on this problem. If you use the op maps, it should also be correct.

**Comment by tonkmr** [comment]
I merged feature 22 into feature 17 to get rid of the problem for the test. Obviously, we wouldn't do that with a real simulation. Hopefully it will pas now.

**Comment by dschwen** [comment]
Previously the GrainTracker found _one too many_ grains, now it is complaining about _missing_ one...

**Comment by tonkmr** [comment]
I have cleaned up the data file more, and which will hopefully help.

**Comment by tonkmr** [comment]
@dschwen and @permcody I hope to eventually get this data file to work. However, it is clear from these issues that the interaction between EBSDReader and GrainTracker is very sensitive to the data file. In the future, we should probably try to come up with a way to make it less sensitive.

Possible issues with data files could include:
- Split grains
- Small grains (just a few pixels)

**Comment by dschwen** [comment]
I'm thinking we should probably add the -1 and +1 grain cases _both_ as _expected faluer_ tests so that we can keep a close eye on GrainTracker behavioral changes.

**Comment by dschwen** [comment]
*snicker*

**Comment by tonkmr** [comment]
I think I have figured out the problem here. The EBSDReader counts the void region as a grain but the GrainTracker does not, so the grain numbers are mismatched. So, EBSDReader needs to be modified so it doesn't count the void region as a grain.

**Comment by dschwen** [comment]
Ok, I think I can do that.

**Comment by permcody** [comment]
Hmm... I really need to continue working on robustness of the EBSD Reader
for these one off cases. Instead of looking at nearest centroids for
instance (for matching up EBSD data to FeatureFloodCount data). I need to
just ask the reader for the number. That I can do.

**Comment by permcody** [comment]
I've completely rewritten how the EBSD tracking works. The new version does a flood based on the EBSD Reader information instead OP information. This should be a lot more reliable than the existing algorithm that tries to match up grain centroids to EBSD centroids. The only problem is it's complicated. My current set of commits is starting to add up to a lot of new logic but I'm close. Everything looks good in serial but I'm still running into a few problems on the parallel runs. Expect a PR soon!

**Comment by permcody** [comment]
These commits and several modes appear in #7605.

**Pull request closed without merging**